Updates to SMELLIE run structure and ELLIE GUI's

EdLeming opened the pull request:
This is quite a large, as yet incomplete, pull request which addresses a lot of issues. I began working this way as Tony had suggested after the last dark running that single (large!) request was his preference for reviewing. After a chat at the last collab meeting I'm not sure if this is still true. 

Freija mentioned in her e-mail for the coming DAQ call that we should discuss how to merge new code. I'm therefore making this available now incase it is of interest for that discussion.

tlatorre-uchicago commented on Source/Experiments/SNOP/SNOPController.m:
```diff
@@ -938,65 +941,63 @@ - (id)comboBox:(NSComboBox *)aComboBox objectValueForItemAtIndex:(NSInteger)inde
-                [loadedSmellieMinIntensityLaser setStringValue:[smellieRunFile objectForKey:@"min_laser_intensity"]];
-                
+                //[loadedSmellieMaxIntensityLaser setStringValue:[smellieRunFile objectForKey:@"max_laser_intensity"]];
+                //[loadedSmellieMinIntensityLaser setStringValue:[smellieRunFile objectForKey:@"min_laser_intensity"]];
```
Are these lines needed? Or can you delete them?

EdLeming replied:
No, they can go. There's still a couple of warnings I need to addess in the ELLIEModel / Controller and one in the SNOPController. I'll tidy those, remove some other commented out lines and push up.

icoulter commented on Source/Objects/Custom Hardware/SNO+/TUBii/TUBiiModel.m:
```diff
@@ -261,13 +258,13 @@ - (void) fireTelliePulser{
     // Stops the Tellie pulser by setting the number of pulses to be fired to zero.
     [self setTellieNPulses:0];
-    NSString* const command=@"FireTelliePulser";
+    NSString* const command=@"SetTelliePulser 1 0.1 0";
```
Thanks. Don't know how this slipped through. Too much copy & paste maybe.
While you're doing this, could you make similar changes for the Smellie and Generic pulsers? They work the same way so it'll need to be done eventually.

EdLeming replied:
Should be done. Have a look

icoulter replied:
👍

BenLand100 commented on Source/Experiments/SNOP/SNOPController.m:
```diff
@@ -1169,6 +1182,59 @@ - (IBAction) emergencySmellieStopAction:(id)sender
+                          userInfo:nil];
+        [e raise];
+    }
+    ELLIEModel* theELLIEModel = [objs objectAtIndex:0];
```
You might also want to enforce that there's only one ELLIE model at a time, but an administrative control of only including one ELLIE object in the experiment is sufficient for testing.

BenLand100 replied:
After seeing this multiple places I decided this is just part of using Orca so no change needed.

BenLand100 commented on Source/Objects/Custom Hardware/SNO+/ELLIE/ELLIEController.m:
```diff
@@ -31,53 +37,60 @@ -(id)init
+            configForSmellie = [[anELLIEModel fetchConfigurationFile:
+                                    [anELLIEModel fetchRecentConfigVersion]] mutableCopy];
+        });
+        dispatch_release(smellieLoadQueue);
```
This bothers me a little - you may not want to release the queue immediately as presumably the async operation might not be done.  Not sure what the objective-c broiler plate could should be though (this may be perfectly fine).

EdLeming replied:
My understanding was this is ok due to objective c reference counting. 

http://stackoverflow.com/questions/5510765/what-is-the-correct-way-to-release-a-gcd-dispatch-queue-property

"Any pending blocks submitted to a queue hold a reference to that queue, so the queue is not deallocated until all pending blocks have completed."

BenLand100 replied:
I think that only applies when ARC (automatic reference counting) is enabled - unfortunately almost all objective c resources/answers assume people are using it (because ARC is great) but Orca is pre-ARC. (Again, I'm not sure it's wrong but I have some doubts.)

BenLand100 commented on Source/Objects/Custom Hardware/SNO+/ELLIE/ELLIEController.m:
```diff
@@ -127,19 +145,96 @@ - (void) registerNotificationObservers
+    dispatch_async(tellieLoadQueue, ^{
+        [model loadTELLIEStaticsFromDB];
+    });
+    dispatch_release(tellieLoadQueue);
```
Same as other similar form. Maybe consider using one of the global concurrent dispatch queues instead? https://developer.apple.com/library/content/documentation/General/Conceptual/ConcurrencyProgrammingGuide/OperationQueues/OperationQueues.html

BenLand100 commented on Source/Objects/Custom Hardware/SNO+/ELLIE/ELLIEModel.m:
```diff
@@ -363,27 +574,79 @@ -(void) fireTellieFibreMaster:(NSMutableDictionary*)fireCommands
-
-    //TELLIE pin readout is an average measurement of the passed "number_of_shots". If a large number of shots are requested
-    //it is useful to split the data into smaller chunks in order to get multiple pin readings.
+    ORRunModel* runControl = [runModels objectAtIndex:0];
```
Similar for here and above - may want to e.g. log if there are more than 1 of each object in the experiment because strange behavior could ensue.

BenLand100 commented on Source/Objects/Custom Hardware/SNO+/ELLIE/ELLIEModel.m:
```diff
@@ -363,27 +574,79 @@ -(void) fireTellieFibreMaster:(NSMutableDictionary*)fireCommands
+    }else{
+        //Stop the current run and start a new run
+        NSLogColor([NSColor redColor], @"[SMELLIE]: Restarting run! \n");
+        [runControl performSelectorOnMainThread:@selector(restartRun) withObject:nil waitUntilDone:YES];
```
Will ultimately want to check with @tlatorre-uchicago to see if this is OK but it seems fine to me.

BenLand100 commented on Source/Objects/Custom Hardware/SNO+/ELLIE/ELLIEModel.m:
```diff
@@ -416,59 +696,181 @@ -(void) fireTellieFibreMaster:(NSMutableDictionary*)fireCommands
+    });
+    return;
+
+err:
```
Make the following block its own selector (method) and change the `goto` to above to a selector call and a return statement. https://xkcd.com/292/

BenLand100 commented on Source/Objects/Custom Hardware/SNO+/ELLIE/ELLIEModel.m:
```diff
@@ -416,59 +696,181 @@ -(void) fireTellieFibreMaster:(NSMutableDictionary*)fireCommands
+                          exceptionWithName:@"noRunModel"
+                          reason:@"*** Please add a ORRunModel to the experiment"
+                          userInfo:nil];
+        NSLogColor([NSColor redColor], @"[TELLIE]: Couldn't find ORRunModel %@\n", [e reason]);
```
Avoid creating the NSException object here too

BenLand100 commented on Source/Objects/Custom Hardware/SNO+/ELLIE/ELLIEModel.m:
```diff
@@ -489,28 +891,29 @@ -(void) pushInitialTellieRunDocument
                           reason:@"*** Please add a ORRunModel to the experiment"
                           userInfo:nil];
-        [e raise];
+        NSLogColor([NSColor redColor], @"[TELLIE]: Couldn't find ORRunModel %@\n", [e reason]);
```
Same

BenLand100 commented on Source/Objects/Custom Hardware/SNO+/ELLIE/ELLIEModel.m:
```diff
@@ -522,20 +925,33 @@ -(void) pushInitialTellieRunDocument
+                          exceptionWithName:@"noRunModel"
+                          reason:@"*** Please add a ORRunModel to the experiment"
+                          userInfo:nil];
+        NSLogColor([NSColor redColor], @"[TELLIE] Couldn't find ORRunModel %@\n", [e reason]);
```
Same

BenLand100 commented on Source/Objects/Custom Hardware/SNO+/ELLIE/ELLIEModel.m:
```diff
@@ -575,18 +991,18 @@ -(void) loadTELLIEStaticsFromDB
                           reason:@"*** Please add a SNOPModel to the experiment"
                           userInfo:nil];
-        [e raise];
+        NSLogColor([NSColor redColor], @"[TELLIE]: STATICS NOT LOADED %@\n", [e reason]);
```
Same

BenLand100 commented on Source/Objects/Custom Hardware/SNO+/ELLIE/ELLIEModel.m:
```diff
@@ -831,177 +1317,118 @@ -(NSArray*)getSmellieRunFrequencyArray:(NSDictionary*)smellieSettings
-     method. I'll look into doing this after the DAQ meeting in Jan - Once I know the tellie
-     one works!
-     */
+     Form a smellie run using the passed smellie run file, stored in smellieSettings dictionary.
```
For this method similarly get rid of the goto's and check with Tony to make sure the run control notifications behave as expected.

BenLand100 commented on Source/Experiments/SNOP/SNOPModel.h:
```diff
@@ -286,8 +285,8 @@
-- (void) getSmellieRunListInfo;
-- (NSMutableDictionary*)smellieTestFct;
+-(void) getSmellieRunFiles;
+-(void) testerHv;
```
Didn't see this implemented anywhere - if it's no used the declaration should be removed.

BenLand100 commented:
Generally speaking there's nothing in this PR that would make it unfit for testing on the detector (provided we're can validate it doesn't blast light into the detector prior to turning on HV). There are some minor code-style issues and a few things to check are consistent with run control behavior, but otherwise it looks good. I didn't check any of the ELLIE specific logic in detail of course but I assume that will all be validated in testing.

EdLeming commented:
Hey @BenLand100. I've addressed the comments you made, have a look. On @tlatorre-uchicago's recommendation I took out all run level control. Subrun control is still included.

BenLand100 commented:
@EdLeming there are still some `goto`s in the ELLIEModel.m file that ideally should be flow control statements and method calls instead of `goto` but I have strong opinions against gotos and am willing to hear other opinions on the matter. Last things to check are:
* Does this compile on the old xcode on the teststand?
* Can you squash the commits into a single commit with a message:
```
SNO+: [short message]

[description]

Pull request by Ed Leming
```
* Looks like master was merged in so that's good to go.

BenLand100 commented:
Looks good. Let me know the outcome of the testing and we can merge this in.

EdLeming commented:
Hey @BenLand100 . Thanks for your help. We spotted a few issues to be resolved (I'll look into it tomorrow). I actually inherited the goto: err: from Tony. It exists a few other places in the code base already.  We can talk about it again once I tidy these issues.

EdLeming commented:
Hey @BenLand100, I think this is good to go. Could you have a look over it again? Once it's got the OK I'll squash it all down.

BenLand100 commented:
Sorry for the delay - this all looks good to me. There's actually no need to squash it down. I can do that during the merge. Let me just check that it builds on an older xcode and I'll go ahead and merge this into master.

BenLand100 commented:
Seems ORCA code unrelated to SNO+ no longer compiles in old xcode, so last thing I'll need from you is a commit message.

EdLeming commented:
Great! Thanks for looking over it again. Commit message: 

"I number of updates to the ELLIE fire sequence structure including the integration of TUBii. Additional updates to the TELLIE GUI to require multiple step validation of user input."

BenLand100 commented:
Merged. FYI - there was a minor issue merging with master. Javi changed the include file with run type constants and the names of those constants. It was unambiguous to fix (three lines) so I went ahead and made the change. You can diff master with your branch to see it, but I would fork off current master for future changes instead of continuing this branch.

EdLeming commented:
👍
